### PR TITLE
manual port of unique ptr pr

### PR DESCRIPTION
**Pull request by wjwwood**
manual port of https://github.com/ros/class_loader/pull/38

This was painful 😄. I think `class_loader` needs to be brought closer to the ROS 1 version as I've been trying to do with `pluginlib`, but for now this lets me move forward.

### CODE REVIEW

**Comment by wjwwood** [comment]
Travis is failing due to poco not building with an error about the c++14 flag. I'll run ci.ros2.org CI instead.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/3394/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/639/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2723/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/3441/)

**Comment by wjwwood** [comment]
Finally figured out the issue on Windows (I think).

**Comment by wjwwood** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/3397/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/642/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2726/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/3444/)

**Comment by wjwwood** [comment]
@mikaelarguedas I still don't know about the travis ci, but the other CI is green. Can you re-review?

**Comment by mikaelarguedas** [comment]
Yep will do, not sure I'll have time to look at it today though.

>  I still don't know about the travis ci

Yeah it appeared on other PRs too (some discussion [here](https://github.com/ros/class_loader/pull/48#issuecomment-337695965)), I didnt have time to look into it but it's not due to this change :+1:

File: include/class_loader/multi_library_class_loader.h [hunk 1]
```diff
@@ -235,6 +278,27 @@ class CLASS_LOADER_PUBLIC MultiLibraryClassLoader
...
+    for (ClassLoaderVector::iterator i = loaders.begin(); i != loaders.end(); ++i)
+    {
+      if (!(*i)->isLibraryLoaded()) {
+        (*i)->loadLibrary();
+      }
+      if ((*i)->isClassAvailable<Base>(class_name)) {
+        return *i;
+      }
+    }
+    return nullptr;
```
Comment (mikaelarguedas) [inline include/class_loader/multi_library_class_loader.h]: looks like sometime you use NULL to match the ROS1 version and sometimes nullptr. Any reason this ones uses another one than the one from upstream?

**Review by mikaelarguedas — COMMENTED** [review]
Thanks @wjwwood for the painful port! (I feel you), I may have one or two comments so I'll wait until tomorrow to finish reviewing and merging.

File: include/class_loader/multi_library_class_loader.h (hunk 1 above)
Comment (wjwwood) [reply include/class_loader/multi_library_class_loader.h]: That's just habit (to use `nullptr`). It's better to use than `NULL`, but also C++11 only:

http://en.cppreference.com/w/cpp/language/nullptr

I'll replace this with `NULL` for now, so it's closer to the ROS 1 style.

Comment (wjwwood) [reply include/class_loader/multi_library_class_loader.h]: I left two which are within a c++11 block and so they can use `nullptr` safely.

File: include/class_loader/class_loader.h [hunk 2]
```diff
@@ -100,54 +108,62 @@ class ClassLoader
...
   }
+#endif
 
+  /// Generates an instance of loadable classes (i.e. class_loader).
   /**
-   * @brief  Generates an instance of loadable classes (i.e. class_loader). It is not necessary for the user to call loadLibrary() as it will be invoked automatically if the library is not yet loaded (which typically happens when in "On Demand Load/Unload" mode).
-   * @param derived_class_name The name of the class we want to create (@see getAvailableClasses())
+   * It is not necessary for the user to call loadLibrary() as it will be
+   * invoked automatically if the library is not yet loaded (which typically
+   * happens when in "On Demand Load/Unload" mode).
```
Comment (mikaelarguedas) [inline include/class_loader/class_loader.h:157]: looking at the upstream PR, some comment content has not been propagated:
```
     * Creating an unmanaged instance disables dynamically unloading libraries when
     * managed pointers go out of scope for all class loaders in this process.
```

File: include/class_loader/multi_library_class_loader.h [hunk 3]
```diff
@@ -100,15 +97,63 @@ class CLASS_LOADER_PUBLIC MultiLibraryClassLoader
...
+        "MultiLibraryClassLoader: Could not create object of class type " + class_name +
+        " as no factory exists for it. "
+        "Make sure that the library exists and was explicitly loaded through "
+        "MultiLibraryClassLoader::loadLibrary()");
+    }
+    return loader->createUniqueInstance<Base>(class_name);
+  }
+
+  /**
+   * @brief Creates an instance of an object of given class name with ancestor class Base
```
Comment (mikaelarguedas) [inline include/class_loader/multi_library_class_loader.h]: Nit: all other functions in this PR have the brief outside of the commented block with `///`

File: include/class_loader/class_loader.h [hunk 4]
```diff
@@ -100,54 +108,62 @@ class ClassLoader
...
 
-    std::shared_ptr<Base> smart_obj(
-      obj, std::bind(&class_loader::ClassLoader::onPluginDeletion<Base>, this, std::placeholders::_1));
-    return smart_obj;
+#if __cplusplus >= 201103L
+  /// Generates an instance of loadable classes (i.e. class_loader).
+  /**
+   * It is not necessary for the user to call loadLibrary() as it will be
+   * invoked automatically.
+   * If the library is not yet loaded (which typically happens when in
```
Comment (mikaelarguedas) [inline include/class_loader/class_loader.h]: Nit: I think this is the continuation of the previous sentence and not a new one (same below)

File: include/class_loader/class_loader.h [hunk 5]
```diff
@@ -236,12 +252,66 @@ class ClassLoader
     }
   }
 
+  /// Generates an instance of loadable classes (i.e. class_loader).
+  /**
+   * It is not necessary for the user to call loadLibrary() as it will be
+   * invoked automatically.
+   * if the library is not yet loaded (which typically happens when in
```
Comment (mikaelarguedas) [inline include/class_loader/class_loader.h]: Nit: I think this is the continuation of the previous sentence and not a new one

**Review by mikaelarguedas — APPROVED** [review]
Thanks @wjwwood for iterating. Change looks good to me, I have one question inline about some mismatch in documentation.
I don't mind fixing the nitpicks myself

**Comment by wjwwood** [comment]
No worries, I think I addressed all the nitpicks.

**Comment by mikaelarguedas** [comment]
> No worries, I think I addressed all the nitpicks.

:+1: looks good to me, merging

**Comment by mikaelarguedas** [comment]
@wjwwood FYI: Big oversight on my side, I mostly compared the changes to the ported ROS1 PR but didnt take into account the ROS2 context. This new test require both `catkin` and `boost` so we will need to change this to get them to run in ROS2. 
But first we'll need to reenable previous tests and make them pass before focusing on this one.
I'll track this on https://github.com/ros/class_loader/pull/59

**Comment by wjwwood** [comment]
Oops, my bad too. I didn't even think about it after it compiled and ran some tests which passed :/

**Pull request merged**